evm/src/arithmetic/shift.rs

```diff
@@ -0,0 +1,394 @@
+        let zero = builder.zero_extension();
+        let mut quo = [zero; 2 * N_LIMBS];
+        quo[..N_LIMBS].copy_from_slice(&lv[quo_range]);
+        quo
+    };
+    let rem = read_value(lv, rem_range);
+
+    let mut constr_poly =
+        modular_constr_poly_ext_circuit(lv, nv, builder, yield_constr, filter, rem, den, quo);
+
+    let input = num;
+    pol_sub_assign_ext_circuit(builder, &mut constr_poly, input);
+
+    for &c in constr_poly.iter() {
+        let t = builder.mul_extension(filter, c);
+        yield_constr.constraint_transition(builder, t);
+    }
+}
+
+pub fn eval_ext_circuit<F: RichField + Extendable<D>, const D: usize>(
+    builder: &mut CircuitBuilder<F, D>,
+    lv: &[ExtensionTarget<D>; NUM_ARITH_COLUMNS],
+    nv: &[ExtensionTarget<D>; NUM_ARITH_COLUMNS],
+    yield_constr: &mut RecursiveConstraintConsumer<F, D>,
+) {
+    eval_ext_circuit_shl(builder, lv, yield_constr);
+    eval_ext_circuit_shr(builder, lv, nv, yield_constr);
+}
+
+#[cfg(test)]
+mod tests {
+    use plonky2::field::goldilocks_field::GoldilocksField;
+    use plonky2::field::types::{Field, Sample};
+    use rand::{Rng, SeedableRng};
+    use rand_chacha::ChaCha8Rng;
+
+    use super::*;
+    use crate::arithmetic::columns::NUM_ARITH_COLUMNS;
+    use crate::constraint_consumer::ConstraintConsumer;
+
+    const N_RND_TESTS: usize = 1000;
+
+    // TODO: Should be able to refactor this test to apply to all operations.
+    #[test]
+    fn generate_eval_consistency_not_shl() {
+        type F = GoldilocksField;
+
+        let mut rng = ChaCha8Rng::seed_from_u64(0x6feb51b7ec230f25);
+        let mut lv = [F::default(); NUM_ARITH_COLUMNS].map(|_| F::sample(&mut rng));
+        let nv = [F::default(); NUM_ARITH_COLUMNS].map(|_| F::sample(&mut rng));
+
+        // if `IS_SHL == 0` and ÌS_SHR == 0`, then the constraints should be met even
+        // if all values are garbage.
+        lv[IS_SHL] = F::ZERO;
+        lv[IS_SHR] = F::ZERO;
+
+        let mut constraint_consumer = ConstraintConsumer::new(
+            vec![GoldilocksField(2), GoldilocksField(3), GoldilocksField(5)],
+            GoldilocksField::ONE,
+            GoldilocksField::ONE,
+            GoldilocksField::ONE,
+        );
+        eval_packed_generic(&lv, &nv, &mut constraint_consumer);
+        for &acc in &constraint_consumer.constraint_accs {
+            assert_eq!(acc, GoldilocksField::ZERO);
+        }
+    }
+
+    #[test]
+    fn generate_eval_consistency_shl() {
+        type F = GoldilocksField;
+
+        let mut rng = ChaCha8Rng::seed_from_u64(0x6feb51b7ec230f25);
+        let mut lv = [F::default(); NUM_ARITH_COLUMNS].map(|_| F::sample(&mut rng));
+        let mut nv = [F::default(); NUM_ARITH_COLUMNS].map(|_| F::sample(&mut rng));
+
+        // set `IS_MUL == 1` and ensure all constraints are satisfied.
+        lv[IS_SHL] = F::ONE;
+        lv[IS_SHR] = F::ZERO;
+
+        for _i in 0..N_RND_TESTS {
+            let shift = U256::from(rng.gen::<usize>());
```

Comment:
If `shift` is a random `usize`-sized value, then the probability that it is less than 256 is about 2^-56. :)
```suggestion
            let shift = U256::from(rng.gen::<u8>());
```
Probably better to have a separate test for big shifts where the expected result is zero.